### DISCUSSION TITLE
Relationship loading strategy to enforce the "use_get" optimisation

### DISCUSSION THREAD

**Discussion by karolinepauls** [sqlalchemy/sqlalchemy#10057] [Ideas]
The `use_get` code path is implemented here: https://github.com/sqlalchemy/sqlalchemy/blob/8fc9e38c375bb851792e3521b141edfdf8aa83e1/lib/sqlalchemy/orm/strategies.py#L925

I'm batch importing into 2 related tables by using Postgres UPSERT in SQLAlchemy 1.4 future mode. Let's call them A and B. B is imported after A and points to A. Ideally I'd return a list containing instances of B, with the related A objects loaded from the session, without any having to do eager loads for A objects when upserting B, or lazy loads later, on access to `b.a`.

As long as A objects are in the session, following code accessing `b.a` will use the "use_get" path. If there is no longer a reference to A objects, they disappear from the session and the "use_get" code path fails, which results in lazy loads.

I solve it by adding `.options(immediateload("a"))` when upserting B. This works because they manage to hit the "use_get" optimisation. However, immediateload (and similarly, selectinload) are able to query the database, so the risk is still there.

I would like to have a "sessiongetload" loading strategy which exclusively uses the session, and fails otherwise.

Code sample:
```
    statement = insert(B).values(insert_values)

    results: list[B] = db.session.execute(
        select(B)
        .from_statement(
            statement.on_conflict_do_update(
                index_elements=[B.a_id],
                set_={
                    ...,
                },
            ).returning(
                B,
            )
        )
        .options(
            immediateload("a"),  # Will not send any SQL because the A objects are in the session.
            raiseload("*"),  # Make sure no lazy (N+1) loading can happen.
        )
        .execution_options(populate_existing=True)
    ).all()
```

**Comment by CaselIT** [comment]
Hi,

This seems the raiseload with sql_only=True
https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#sqlalchemy.orm.raiseload.params.sql_only

**Comment by karolinepauls** [comment]
I don't think raiseload will work, since it won't prevent the related object from disappearing from the identity map when it is no longer referenced.

What I want is a loading strategy that is eager but only pulls from the identity map. If the strategy isn't eager, related objects don't get "cached" by the relationship, and may be GC'd when they are no longer referenced in local vars. When that happens, even raiseload with sql_only=True will raise.

**Reply by zzzeek** [reply]
raise with sql_only is the correct option that will raise if SQL were to be emitted.  As far as "disappearing from the identity map", you have to enforce that on your end by creating a strong reference to the object(s) that you want to stay around.   Place the object(s) into a list or a set, and if you prefer you can even associate it with the `Session` by placing it in the [info](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.info) dictionary.

for the use_get kind of case, it's often worth it to do a query up front for all the m2o objects that you'll want:

```py
preload_m2os = session.scalars(select(Related).where(Related.id.in_( ... ids here, either a subquery from the parent object, or from batches of records, etc. ...)).all()
```

**Reply by karolinepauls** [reply]
The whole point is that I don't want to create any additional strong references. Not because I'm lazy, but because I want to write as straightforward code as possible. It's much easier to explain in a docstring that "Returned B objects have related A's attached" than to "We retain strong references to A objects so that accesses to B.a will retrieve related instances of A from the session.". 

The `immediateload("a")`, which manages to load objects from the session, works great, since it attaches related A's to the returned B's, and is easier to use. It would simply be better if it was 100% clear that it can only access the session.

**Comment by zzzeek** [comment]
> The whole point is that I don't want to create any additional strong references.

well sure, neither does SQLAlchemy.      Hence the conundrum here.   SQLAlchemy Session has to be weak referencing.    There used to be a "strong referencing" variant but as you can imagine it was a disaster because it became the source of all kinds of "memory leaks", as people iterated through thousands of objects.    

There's not really a solution to that other than, "the end user needs to indicate **which** objects should not be garbage collected".  Which you do quite simply by creating a strong reference to it.  this is like a one-liner kind of thing.


> Not because I'm lazy, but because I want to write as straightforward code as possible. 

that's fine but then your code is not optimized.  your use case here is a performance optimization.  Performance optimizations are necessarily going to express themselves as small little bumps in one's otherwise completely minimalistic codebase.   It's quite trivial in this case.

> It's much easier to explain in a docstring that "Returned B objects have related A's attached" than to "We retain strong references to A objects so that accesses to B.a will retrieve related instances of A from the session.".

That's fine, pre-load those many-to-ones and use immediateload, there you go.  this is what immediateload is for.

 
> The `immediateload("a")`, which manages to load objects from the session, works great, since it attaches related A's to the returned B's, and is easier to use. It would simply be better if it was 100% clear that it can only access the session.

so an "immediateload but no SQL" version of it.   Might be possible, not something we have right now.
